maintenance/ensembl.py

```diff
@@ -0,0 +1,99 @@
+        self.max_requests_per_second = max_requests_per_second
+        self.num_requests = 0
+        self.last_request_time = 0
+
+    def _make_request(self, endpoint, headers, params):
+        if headers is None:
+            headers = {}
+        else:
+            headers = dict(headers)
+
+        if "Content-type" not in headers:
+            headers["Content-type"] = "Application/json"
+
+        quoted_endpoint = urllib.parse.quote(endpoint)
+        url = urllib.parse.urljoin(self.server, quoted_endpoint)
+        # skipping the params for now
+        assert params is None
+        return urllib.request.Request(url, headers=headers)
+
+    def _sleep_if_needed(self):
+        # check if we need to rate limit ourselves
+        if self.num_requests >= self.max_requests_per_second:
+            delta = time.time() - self.last_request_time
+            if delta < 1:
+                logging.debug("Rate limiting REST API")
+                time.sleep(1 - delta)
+            self.num_requests = 0
+        else:
+            self.num_requests += 1
+        self.last_request_time = time.time()
+
+    def get(self, endpoint, headers=None, params=None):
+        """
+        Runs a HTTP get request to the specified endpoint with the
+        specified headers and parameters, and returns a dictionary
+        consisting of the response.
+        """
+        self._sleep_if_needed()
+        request = self._make_request(endpoint, headers, params)
+        logging.debug("making request to %s", request.full_url)
+        response = urllib.request.urlopen(request)
+        content = response.read()
+        data = json.loads(content)
+        return data
+
+    def get_release(self):
+        """
+        Returns the current ensembl release number.
+        """
+        output = self.get(endpoint=f"/info/data")
+        releases = output["releases"]
+        # Docs say: may return more than one release
+        # (unfrequent non-standard Ensembl configuration).
+        assert len(releases) == 1
+        return releases[0]
+
+    def get_genome_data(self, ensembl_id):
+        """
+        Returns the genome data for the specified Ensembl species
+        ID (e.g., "homo_sapiens"). This is a subset of what
+        is returned by the Ensembl API, to restrict to the information
+        that we're interested in for stdpopsim.
+        """
+        output = self.get(endpoint=f"/info/assembly/{ensembl_id}")
+        data = {
+            "assembly_accession": output["assembly_accession"],
+            "assembly_name": output["assembly_name"],
+        }
+        chromosomes = {}
+        for region in output["top_level_region"]:
+            if region["coord_system"] == "chromosome":
+                chromosomes[region["name"]] = {"length": region["length"]}
+        # Make sure the chromosomes are sorted by name
+        data["chromosomes"] = {
+            name: chromosomes[name] for name in sorted(chromosomes.keys())}
```

Comment:
I think the "karyotype" output field gives a more natural ordering than sorting.
```
>>> sorted(["1", "2", "10"])
['1', '10', '2']
```

Reply:
Yes, excellent idea. I'll change that.